libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -0,0 +1,232 @@
+	)X");
+	t("object", runtimeObjectName(_contract));
+	return t.render();
+}
+
+string IRGenerator::callValueCheck()
+{
+	return "if callvalue() { revert(0, 0) }";
+}
+
+string IRGenerator::creationObjectName(ContractDefinition const& _contract)
+{
+	return _contract.name() + "_" + to_string(_contract.id());
+}
+
+string IRGenerator::runtimeObjectName(ContractDefinition const& _contract)
+{
+	return _contract.name() + "_" + to_string(_contract.id()) + "_deployed";
+}
+
+string IRGenerator::dispatchRoutine(ContractDefinition const& _contract)
+{
+	Whiskers t(R"X(
+		if iszero(lt(calldatasize(), 4))
+		{
+			let selector := <shr224>(calldataload(0))
+			switch selector
+			<#cases>
+			case <functionSelector>
+			{
+				// <functionName>
+				<callValueCheck>
+				<assignToParams> <abiDecode>(4, calldatasize())
+				<assignToRetParams> <function>(<params>)
+				let memPos := <allocate>(0)
+				let memEnd := <abiEncode>(memPos <comma> <retParams>)
+				return(memPos, sub(memEnd, memPos))
+			}
+			</cases>
+			default {}
+		}
+		<fallback>
+	)X");
+	t("shr224", m_utils.shiftRightFunction(224));
+	vector<map<string, string>> functions;
+	for (auto const& function: _contract.interfaceFunctions())
+	{
+		functions.push_back({});
+		map<string, string>& templ = functions.back();
+		templ["functionSelector"] = "0x" + function.first.hex();
+		FunctionTypePointer const& type = function.second;
+		templ["functionName"] = type->externalSignature();
+		templ["callValueCheck"] = type->isPayable() ? "" : callValueCheck();
+
+		unsigned paramVars = make_shared<TupleType>(type->parameterTypes())->sizeOnStack();
+		unsigned retVars = make_shared<TupleType>(type->returnParameterTypes())->sizeOnStack();
+		templ["assignToParams"] = paramVars == 0 ? "" : "let " + m_utils.suffixedVariableNameList("param_", 0, paramVars) + " := ";
+		templ["assignToRetParams"] = retVars == 0 ? "" : "let " + m_utils.suffixedVariableNameList("ret_", 0, retVars) + " := ";
+
+		ABIFunctions abiFunctions(m_evmVersion, m_context.functionCollector());
+		templ["abiDecode"] = abiFunctions.tupleDecoder(type->parameterTypes());
+		templ["params"] = m_utils.suffixedVariableNameList("param_", 0, paramVars);
+		templ["retParams"] = m_utils.suffixedVariableNameList("ret_", retVars, 0);
+		templ["function"] = generateIRFunction(dynamic_cast<FunctionDefinition const&>(type->declaration()));
+		templ["allocate"] = m_utils.allocationFunction();
+		templ["abiEncode"] = abiFunctions.tupleEncoder(type->returnParameterTypes(), type->returnParameterTypes(), false);
+		templ["comma"] = retVars == 0 ? "" : ", ";
+	}
+	t("cases", functions);
+	if (FunctionDefinition const* fallback = _contract.fallbackFunction())
+	{
+		string fallbackCode;
+		if (!fallback->isPayable())
+			fallbackCode += callValueCheck();
+		fallbackCode += generateIRFunction(*fallback) + "() stop()";
+
+		t("fallback", fallbackCode);
+	}
+	else
+		t("fallback", "revert(0, 0)");
+	return t.render();
+}
+
+string IRGenerator::memoryInit()
+{
```

Comment:
Maybe a comment about memory being zero-initialized and not having to write to the zero pointer wouldn't hurt.